src/compiler/mixin.cc

```diff
@@ -23,6 +23,171 @@
+
+  // Make `visit` public.
+  ir::Node* visit(ir::Node* node) override {
+    return ReplacingVisitor::visit(node);
+  }
+
+ protected:
+  ir::Node* visit_Super(ir::Super* node) override {
+    if (node->expression() == null) {
+      auto replacement = visit_static_super_call(null, node->range());
+      if (replacement != null) {
+        node->replace_expression(replacement->as_Expression());
+      }
+      return node;
+    }
+    in_super_ = true;
+    auto result = ir::ReplacingVisitor::visit_Super(node);
+    in_super_ = false;
+    has_seen_super_ = true;
+    return result;
+  }
+
+  ir::Node* visit_CallStatic(ir::CallStatic* node) override {
+    if (!in_super_) return ir::ReplacingVisitor::visit_CallStatic(node);
+    auto method = node->target()->target();
+    if (method->is_constructor() && method->holder() == super_) {
+      return visit_static_super_call(node, node->range());
+    }
+    return ir::ReplacingVisitor::visit_CallStatic(node);
+  }
+
+  /// May be null if there wasn't any explicit call.
+  virtual ir::Node* visit_static_super_call(ir::CallStatic* node, const Source::Range& range) = 0;
+
+ private:
+  ir::Class* super_;
+  bool in_super_ = false;
+  bool has_seen_super_ = false;
+};
+
+class MixinConstructorVisitor : protected SuperCallVisitor {
+ public:
+  MixinConstructorVisitor(ir::Class* holder, List<ir::Field*> fields)
+      : SuperCallVisitor(holder)
+      , fields_(fields) {}
+
+  bool has_seen_super() const { return has_seen_super_; }
+
+  // Make `visit` public.
+  ir::Node* visit(ir::Node* node) override {
+    return SuperCallVisitor::visit(node);
+  }
+
+ protected:
+  ir::Node* visit_Method(ir::Method* node) override {
+    ASSERT(node->is_constructor());
+    if (!fields_.is_empty()) {
+      ListBuilder<ir::Expression*> new_body;
+      for (auto field : fields_) {
+        auto range = field->range();
+        bool is_final, is_block;
+        auto local = _new ir::Local(field->name(),
+                                    is_final=false,
+                                    is_block=false,
+                                    field->type(),
+                                    range);
+        field_to_local_[field] = local;
+        // TODO(florian): can we avoid the `null` assignment?
+        auto definition = _new ir::AssignmentDefine(local, _new ir::LiteralNull(range), range);
+        new_body.add(definition);
+      }
+      new_body.add(node->body());
+      node->replace_body(_new ir::Sequence(new_body.build(), node->range()));
+    }
+    ASSERT(node->parameters().length() == 1 && node->parameters()[0]->index() == 0);
+    bool is_block, default_value;
+    int block_index = 1;
+    this_ = node->parameters()[0];
+    next_super_ = _new ir::Parameter(Symbol::synthetic("<next-super>"),
+                                     ir::Type::any(),
+                                     is_block=true,
+                                     block_index,
```

Comment:
Why don't you use the same style for this as for `is_block`? (assign as part of the call)